### ISSUE TITLE
Remove credentials from error responses

### ISSUE THREAD

**Issue by glynnbird** [cloudant/nodejs-cloudant#85]
The `err` variable returned in the callback function contains the Cloudant login credentials. Could we modify the `relax` function in Nano to remove this?

**Comment by glynnbird** [comment]
Can we explore adding credentials into a header rather than a URL to fix this?

**Comment by glynnbird** [comment]
Does passing credentials in as headers (cookie, or otherwise), prevent the error containing auth information?

**Comment by AndrewHuffman** [comment]
Well, basic auth is sent through the `Authorization` header already. It just base64 encodes the `<user>:<pass>.` It doesn't need to be in the URL, though so maybe it depends on how the `request` API is called.

**Comment by AndrewHuffman** [comment]
It does appear the headers are included in the error:

```
        "msg": {
            "message": "{error,timeout}",
            "stack": "Error: {error,timeout}\n    at Request._callback (/home/vcap/app/node_modules/cloudant/lib/nano.js:240:15)\n    at Request.self.callback (/home/vcap/app/node_modules/request/request.js:354:22)\n    at Request.emit (events.js:98:17)\n    at Request.<anonymous> (/home/vcap/app/node_modules/request/request.js:1207:14)\n    at Request.emit (events.js:117:20)\n    at IncomingMessage.<anonymous> (/home/vcap/app/node_modules/request/request.js:1153:12)\n    at IncomingMessage.emit (events.js:117:20)\n    at IncomingMessage.wrapped (/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:157:28)\n    at IncomingMessage.wrappedEmit [as emit] (/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:197:46)\n    at _stream_readable.js:944:16\n    at wrapped (/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:157:28)\n    at wrapped (/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:157:28)\n    at process._tickDomainCallback (node.js:492:13)\n    at process.wrappedFunction (/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:262:51)",
            "name": "Error",
            "error": "badmatch",
            "reason": "{error,timeout}",
            "ref": 441380633,
            "scope": "couch",
            "statusCode": 500,
            "request": {
                "method": "POST",
                "headers": {
                    "content-type": "application/json",
                    "accept": "application/json"
                },
                "uri": "https://xxxxxxxxxxxxxxxxx@XXXXXXXXXXX.cloudant.com/tnt_rlcm0pcx_realtime_event_data_raw/_find",
                "body": "{\"use_index\":[\"_design/visitsAnalytics\",\"site\"],\"selector\":{\"org_code\":{\"$eq\":\"8be108jl\"},\"site_code\":{\"$eq\":\"bo1j0tdd\"},\"registered\":{\"$eq\":false},\"state\":{\"$eq\":1},\"detected_timestamp\":{\"$and\":[{\"$gte\":1439426076412},{\"$lte\":1439512476412}]}},\"fields\":[\"zone_code\",\"zone.name\",\"floor_code\",\"floor.name\",\"site_code\",\"site.name\",\"detected_timestamp\"]}"
            },
            "headers": {
                "x-couch-request-id": "f3e8d47a77",
                "x-cloudant-stack-hash": "441380633",
                "date": "Fri, 14 Aug 2015 00:43:21 GMT",
                "content-type": "application/json",
                "cache-control": "must-revalidate",
                "strict-transport-security": "max-age=31536000",
                "x-content-type-options": "nosniff;",
                "statusCode": 500,
                "uri": "https://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx@presenceinsightsyp.cloudant.com/tnt_rlcm0pcx_realtime_event_data_raw/_find"
            },
            "errid": "non_200",
            "description": "couch returned 500"
        }
```

**Comment by jhs** [comment]
Rather than omit the headers, perhaps scrub them? I have done that in the past to keep passwords out of log files.

An advantage of that technique is that it also makes us safe during logging. I think users could currently hook into our log/debug callbacks and inadvertently log their credentials. (Arguably that is their problem; however considering how much is at stake to lose control of cloud database credentials, I'd argue we should never let such secrets "leak" from this package.)

The disadvantage though is that there is a higher risk of doing it wrong. We'd need to scrub URLs with credential prefixes, "Authorization" headers, probably cookie headers, etc. and we'd want to be reasonably sure we did it correctly.

**Comment by glynnbird** [comment]
I've had a go at scrubbing headers at the Nano level. See this branch of CouchDB Nano:

https://github.com/glynnbird/couchdb-nano/tree/scrubauth

Which contains the following commit:

https://github.com/glynnbird/couchdb-nano/commit/0f7752d7e247fd55b79a7c0831596918487d1978

It fails the Nano project's pre-commit-hook criteria, because there is now less than 100% test coverage, but I find the Nano test framework baffling tbh.

**Comment by AndrewHuffman** [comment]
@glynnbird Do you know which tests are failing on code coverage? I may be able to take a look later this evening.

**Comment by glynnbird** [comment]
@AndrewHuffman I presume it was the code that I wrote ;) although the test coverage may not have been 100% before I started  i.e. a scrub function and a couple of calls to it. https://github.com/glynnbird/couchdb-nano/commit/0f7752d7e247fd55b79a7c0831596918487d1978

**Comment by AndrewHuffman** [comment]
@glynnbird What's your policy for managing upstream defects? Wait for a pull request or fork?

**Comment by glynnbird** [comment]
@AndrewHuffman As Apache has taken over the `nano` project, I would suggest forking https://github.com/apache/couchdb-nano and submitting a PR back. My colleague @jhs is an Apache Committer and should be able to help get any changes merged in.

**Comment by jhs** [comment]
Merged into Nano; ready to push. Still no unit tests however.

**Comment by jhs** [comment]
(Quickly re-added [Agenda] to remember to discuss this next time)

**Comment by tannerlinsley** [comment]
Would be great to get this pushed to nano and updated here.

**Comment by glynnbird** [comment]
That's exactly what's going to happen @tannerlinsley. The changes have been made by @jhs to the couchdb/nano project. When that's published, we'll pick up the dependency here.

**Comment by tannerlinsley** [comment]
Nice!

**Comment by jhs** [comment]
Fixed in v1.4.0
